code/modules/projectiles/guns/magnetic/bore.dm

```diff
@@ -77,91 +74,169 @@
 
-		var/can_hold_val = 0
-		while(can_hold_val < round(max_mat_storage / 2000))
-			if(mat_storage + 2000 <= max_mat_storage && do_after(user,1.5 SECONDS))
+			var/can_hold_val = 0
+			loading = TRUE
+			while(mat_storage + SHEET_MATERIAL_AMOUNT <= max_mat_storage && do_after(user,1.5 SECONDS))
 				can_hold_val ++
-				mat_storage += 2000
+				mat_storage += SHEET_MATERIAL_AMOUNT
 				playsound(src, 'sound/effects/phasein.ogg', 15, 1)
-			else
-				loading = FALSE
-				break
-		M.use(can_hold_val)
+			M.use(can_hold_val)
+			loading = FALSE
+
+		else //ore
+			if(M.material != ammo_material)
+				return
+
+			if(mat_storage + (SHEET_MATERIAL_AMOUNT/2*0.8) > max_mat_storage)
+				to_chat(user, "<span class='warning'>\The [src] cannot hold more [ammo_material].</span>")
+				return
+
+			qdel(M)
+			mat_storage += (SHEET_MATERIAL_AMOUNT/2*0.8) //two plasma ores needed per sheet, some inefficiency for not using refined product
+
 
 		user.visible_message("<span class='notice'>\The [user] loads \the [src] with \the [M].</span>")
 		playsound(src, 'sound/weapons/flipblade.ogg', 50, 1)
 		update_icon()
 		return
+
+	. = ..()
+
+#define GEN_STARTING -1
+#define GEN_OFF 0
+#define GEN_IDLE 1
+#define GEN_ACTIVE 2
+
+/obj/item/weapon/gun/magnetic/matfed/phoronbore
+	name = "portable phoron bore"
+	desc = "A large man-portable tunnel bore, using phorogenic plasma blasts. Point away from user."
+	description_fluff = "An aging Grayson Manufactories mining tool used for rapidly digging through rock. Mass production was discontinued when many of the devices were stolen and used to break into a high security facility by Boiling Point drones."
+	description_antag = "This device is exceptional at breaking down walls, though it is incredibly loud when doing so."
+	description_info = "The projectile of this tool will travel six tiles before dissipating, excavating mineral walls as it does so. It can be reloaded with phoron sheets."
+
+	icon_state = "bore"
+	item_state = "bore"
+	wielded_item_state = "bore-wielded"
+	one_handed_penalty = 5
+
+	projectile_type = /obj/item/projectile/bullet/magnetic/bore
+
+	gun_unreliable = 0
+	power_cost = 100
+	ammo_material = MAT_PHORON
+
+	action_button_name = "Toggle internal generator"
+
+	var/generator_state = GEN_OFF
+	var/datum/looping_sound/small_motor/soundloop
+	var/time_started //to keep the soundloop from being "stopped" too soon and playing indefinitely
+
+/obj/item/weapon/gun/magnetic/matfed/phoronbore/Initialize()
+	. = ..()
+	soundloop = new(list(src), 0)
+
+/obj/item/weapon/gun/magnetic/matfed/phoronbore/Destroy()
+	QDEL_NULL(soundloop)
 	. = ..()
+
+/obj/item/weapon/gun/magnetic/matfed/phoronbore/ui_action_click()
+	toggle_generator(usr)
+
+/obj/item/weapon/gun/magnetic/matfed/phoronbore/process()
+	if(generator_state && !mat_storage)
+		audible_message(SPAN_NOTICE("\The [src] goes quiet."),SPAN_NOTICE("A motor noise cuts out."))
+		soundloop.stop()
+		generator_state = GEN_OFF
+
+	else if(generator_state)
```

Comment:
This can try to set the soundloop speed and generate power while the generator is being started.

Reply:
Whoops, the starting state was a late change and I missed this